Add state attribute to StorageVolume

allardhoeve opened the pull request:
1. Currently, volume state is defined by the driver in a key in `extra`. This key varies between drivers. Also, the value varies between drivers.
2. However, it is a commonly accessed attribute that should be unified across drivers, just as `Node` state is unified in `NodeState`.
3. This PR does exactly that.

Some caveats:
1. Except for OS and EC2, all other drivers return StorageVolume objects with state `UNKNOWN`.
   1. I have no access to GCE.
   2. I could find no exhaustive list of GCE disk states.
   3. The same holds for all the other drivers.
   4. I do not think this is a problem, because the instance variable does not currently exist at all :wink: 

UPDATE: GCE list is here: https://cloud.google.com/compute/docs/reference/latest/disks#resource

```
Acceptable values are:
"CREATING": The disk is being created.
"FAILED": The disk failed to be created.
"READY": The disk is ready to use.
"RESTORING": The disk is being restored from either a snapshot or an image.
```

Commit:
Add state to StorageVolume

closes #476

gertjanol commented:
[EC2 Volume states](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Volume.html): `creating | available | in-use | deleting | deleted | error`
[OpenStack Volume states](http://developer.openstack.org/api-ref-blockstorage-v2.html#volumes-v2): `creating | available | attaching | in-use | deleting | error | error_deleting | backing-up | restoring-backup | error_restoring | error_extending`

allardhoeve commented:
Where did you find this?

allardhoeve commented:
Fixed unknown node states, thanks @gertjanol

Commit:
Document functions and update signature

- Document create_volume_snapshot arguments properly
- Document destroy_volume_snapshot arguments properly
- Update signature to match all existing implementations
  - name should be optional

closes #478

allardhoeve commented:
@Kami, does this need a proposal @ dev?

Kami commented on libcloud/compute/drivers/ec2.py:
```diff
@@ -4683,6 +4695,14 @@ def _to_volume(self, element, name=None):
+                             namespace=NAMESPACE)
+
+        try:
+            state = self.VOLUME_STATE_MAP[raw_state]
```
Minor style thing - doing `self.VOLUME_STATE_MAP.get(raw_state, StorageVolumeState.UNKNOWN)` saves a couple of lines of code. On top of that, code also won't break and require an immediate update if a provider adds a new state (unlikely, but still).

allardhoeve replied:
Done.

Kami commented on libcloud/compute/types.py:
```diff
@@ -221,6 +221,18 @@ def fromstring(cls, value):
 
 
+class StorageVolumeState(object):
+    UNKNOWN = None
```
Should probably also use a number for `unknown` - that's what other ENUM classes to and this way we can differentiate between "no state" and "state not recognized by libcloud".

allardhoeve replied:
And done.

Kami commented:
Need to update the changelog.

Besides that, LGTM, +1

allardhoeve commented:
thanks!

allardhoeve commented:
Documented the changelog. Will wait for the build, then merge.

Kami commented:
Looks like the build is stuck :/

allardhoeve commented:
Travis is very, very slow lately

Kami commented:
Yeah, build just seems to be stuck. Probably best to just merge since since we don't have admin permissions and we can't restart the build (we could push a "dummy" commit but that's annoying...).
